runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/Exceptions.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries
+
+/**
+ * Indicates some failure happened while retrying.
+ * @param message A message indicating the failure mode.
+ * @param cause An underlying exception which caused this exception.
+ * @param attempts The number of attempts made before this failure was encountered.
+ * @param lastResponse The last response received from the retry strategy before this failure was encountered.
+ * @param lastException The exception caught by the retry strategy before this failure was encountered. Note that the
```

Comment:
suggestion

if there is a way to further distinguish between `cause` and `lastException` for clarity...

Reply:
It's a difficult nuance to convey. `cause` is pretty much the standard exception re-wrapping mechanism that exists throughout Java/Kotlin. `lastException` is more analogous to `lastResponse`...it's the last _output_ we received from the block but it's not necessarily the underlying cause of this exception.

Take for example a retry loop that keeps getting a throttling exception (which is retryable). Eventually, we're going to give up and throw a `TimedOutException` or `TooManyAttemptsException`. In those cases, the most recent retryable exception we got from our code block _isn't the cause_ of the timeout or whatever, it's merely the last thing we saw and may be interesting to code that catches the timeout.

Can you think of a succinct way to capture that in KDoc?

Reply:
Yeah it's a tough nut to crack.  In what cases will both `lastResponse` and `lastException` contain a value?  One minor improvement may be to use the word `previous` rather than `last`.   IMO `last` is somewhat ambiguous as it can mean just the last time something happened, which could be multiple iterations in the past, whereas `previous` to me implies something that only could have occurred in the last iteration.  As I write this it's pretty obvious how subjective this all so :shrug: .

Reply:
> In what cases will both `lastResponse` and `lastException` contain a value?

Under no circumstances will both fields contain a value. The last _result_ from a block execution can only be a response or an exception, not both.

Reply:
In that case would it make sense to represent the combination as a union via a sealed class?

Reply:
We had a union called `Outcome` in the design review but it seemed like popular opinion was against it. I can create a union here but it will conflict conceptually and semantically with [`Result`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/) (which is not a union).

Do you have any suggestion on names for the union and the members?